bagel/tests/test_utility.py

```diff
@@ -543,7 +574,216 @@ def test_failed_json_reading_raises_informative_error(
+                    "schemaKey": "Sex",
+                },
+                "schemaKey": "PhenotypicSession",
+            },
+            {
+                "identifier": "nb:eb57d0c1-fb96-4c04-8c16-1f29f7f40db4",
+                "hasLabel": "ses-02",
+                "hasAge": 35.3,
+                "hasSex": {
+                    "identifier": "snomed:248152002",
+                    "schemaKey": "Sex",
+                },
+                "schemaKey": "PhenotypicSession",
+            },
+            {
+                "identifier": "nb:e67fd08b-9bf9-4ed8-b4cc-d0142cd27789",
+                "hasLabel": "ses-im01",
+                "hasFilePath": "/data/neurobagel/bagel-cli/bids-examples/synthetic/sub-01/ses-01",
+                "hasAcquisition": [
+                    {
+                        "identifier": "nb:5dc2e11e-4f7a-4b0e-9488-843f0a607f4b",
+                        "hasContrastType": {
+                            "identifier": "nidm:T1Weighted",
+                            "schemaKey": "Image",
+                        },
+                        "schemaKey": "Acquisition",
+                    },
+                ],
+                "schemaKey": "ImagingSession",
+            },
+        ],
+        "schemaKey": "Subject",
+    }
+    example_subject = models.Subject(**example_subject_jsonld)
+
+    assert list(
+        dutil.get_subject_imaging_sessions(example_subject).keys()
+    ) == ["ses-im01"]
+
+
+def test_create_completed_pipelines():
+    """
+    Test that completed pipelines for a subject-session are accurately identified
+    based on completion status of all pipeline steps.
+    """
+    sub_ses_data = [
+        [
+            "01",
+            "sub-01",
+            "01",
+            "ses-01",
+            "fmriprep",
+            "20.2.7",
+            "step1",
+            "SUCCESS",
+        ],
+        [
+            "01",
+            "sub-01",
+            "01",
+            "ses-01",
+            "fmriprep",
+            "20.2.7",
+            "step2",
+            "FAIL",
+        ],
+        [
+            "01",
+            "sub-01",
+            "01",
+            "ses-01",
+            "fmriprep",
+            "23.1.3",
+            "default",
+            "SUCCESS",
+        ],
+    ]
+    example_ses_proc_df = pd.DataFrame.from_records(
+        # TODO: Don't hardcode col names?
+        columns=[
+            "participant_id",
+            "bids_participant",
+            "session_id",
+            "bids_session",
+            "pipeline_name",
+            "pipeline_version",
+            "pipeline_step",
+            "status",
+        ],
+        data=sub_ses_data,
+    )
+    completed_pipelines = dutil.create_completed_pipelines(example_ses_proc_df)
+
+    assert len(completed_pipelines) == 1
+    assert (
+        completed_pipelines[0].hasPipelineName.identifier
+        == f"{mappings.NP.pf}:fmriprep"
+    )
+    assert completed_pipelines[0].hasPipelineVersion == "23.1.3"
```

Comment:
as mentioned above, assertion is a little hard to understand here since there are different versions in your table. Once that's cleared up, the assertion might need a little comment to explain what's tested here

Reply:
see reply to your comment above!